### PR TITLE
ERRAI-1064: Add support for Elemental2 API

### PR DESCRIPTION
**Pull request by tiagobento**
On this PR:
- Adding support for Elemental2 API. It is possible to use any Elemental2 class in the same way Errai DOM wrappers were used.
- Bumping GWT version to 2.8.1 so that tests using Elemental2 API work.

### CODE REVIEW

File: errai-common/src/main/java/org/jboss/errai/common/client/dom/TableCell.java [hunk 1]
```diff
@@ -27,7 +27,7 @@
  * @see <a href="https://developer.mozilla.org/en-US/docs/Web/API/HTMLTableCellElement">Web API</a>
  */
 @JsType(isNative = true)
-@Element({"tr", "td"})
+@Element({"th", "td"})
```
Comment (tiagobento) [inline errai-common/src/main/java/org/jboss/errai/common/client/dom/TableCell.java]: Sure. I'll create a separate PR.

File: errai-data-binding/src/main/java/org/jboss/errai/databinding/client/BindableProxyAgent.java [hunk 2]
```diff
@@ -738,17 +746,17 @@ void updateWidgetsAndFireEvents() {
...
-      else if (component instanceof HasText) {
+      } else if (component instanceof HasText) {
         updateComponentValue(newValue, (HasText) component, converter);
-      }
-      else if (component instanceof IsElement || isElement(component)) {
-        final Element element = BoundUtil.asElement(component instanceof IsElement ? ((IsElement) component).getElement() : component);
+      } else if (component instanceof IsElement
+              || component instanceof org.jboss.errai.common.client.api.elemental2.IsElement || isElement(component)) {
+
+        final Element element = BoundUtil.asElement(getComponent(component));
```
Comment (tiagobento) [inline errai-data-binding/src/main/java/org/jboss/errai/databinding/client/BindableProxyAgent.java]: Yes, absolutely.

File: errai-ui/src/main/java/org/jboss/errai/ui/rebind/ioc/element/ElementInjectionBodyGenerator.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    stmts.add(declareFinalVariable(elementVar, elementClass(), elementInitialization()));
+
+    for (final Property property : properties) {
+      stmts.add(loadVariable(elementVar).invoke("setPropertyString", loadLiteral(property.name()),
+              loadLiteral(property.value())));
+    }
+
+    final String retValVar = "retVal";
+
+    stmts.add(declareFinalVariable(retValVar, type, invokeStatic(Js.class, "uncheckedCast", loadVariable(elementVar))));
```
Comment (tiagobento) [inline errai-ui/src/main/java/org/jboss/errai/ui/rebind/ioc/element/ElementInjectionBodyGenerator.java]: Should've been, totally missed it! Changing in a bit.

File: errai-ioc/src/main/java/org/jboss/errai/ioc/Container.gwt.xml [hunk 4]
```diff
@@ -22,6 +22,7 @@
   <inherits name="org.jboss.errai.common.Logging"/>
   <inherits name="javax.enterprise.Support"/>
   <inherits name="com.google.common.collect.Collect"/>
+  <inherits name="elemental2.dom.Dom"/>
```
Comment (tiagobento) [inline errai-ioc/src/main/java/org/jboss/errai/ioc/Container.gwt.xml]: After removing the `uncheckedCast`s from `org.jboss.errai.ioc.rebind.ioc.bootstrapper.TypeFactoryBodyGenerator` this should've been removed as well. Changing in a bit.

**Comment by tiagobento** [comment]
@mbarkley Thanks for the comments. I'll follow your suggestions and update the PR soon :)

**Comment by tiagobento** [comment]
Working on adding tests to errai-data-binding now.

**Comment by tiagobento** [comment]
@mbarkley No problem, just finished adding them. I'm going to update the docs now and probably tomorrow you can take a look at the updates and we can merge the PR :)

**Comment by mbarkley** [comment]
Thanks @tiagobento! Will merge when the build passes.

File: errai-docs/src/main/asciidoc/reference.asciidoc [hunk 5]
```diff
@@ -5904,11 +5908,16 @@ public class MyBean {
...
+The interface [code]+org.jboss.errai.common.client.api.elemental2.IsElement+ is analogous to `IsWidget`. It declares a single method `getElement` for accessing a type that is a UI component with a root element.
 
-The interface [code]+org.jboss.errai.common.client.api.IsElement+ is analogous to `IsWidget`. It declares a single method `getElement` for accessing a type that is a UI component with a root element.
+The interface [code]+org.jboss.errai.ui.client.local.api.elemental2.IsElement+ provides a default implementation of [code]+org.jboss.errai.common.client.api.elemental2.IsElement+ that works for any templated bean. For example:
 
-The interface [code]+org.jboss.errai.ui.client.local.api.IsElement+ provides a default implementation of [code]+org.jboss.errai.common.client.api.IsElement+ that works for any templated bean. For example:
+[NOTE]
+====
+Although using [code]+org.jboss.errai.ui.client.local.api.elemental2.IsElement+ is the recomended approach, [code]+org.jboss.errai.ui.client.local.api.IsElement+ is still supported.
+====
```
Comment (tiagobento) [inline errai-docs/src/main/asciidoc/reference.asciidoc]: @mbarkley I added this note here, do you want it to be moved further up?

Comment (tiagobento) [inline errai-docs/src/main/asciidoc/reference.asciidoc]: Oh, you already saw it :)

**Comment by tiagobento** [comment]
@mbarkley Yes. A few test broke so I changed the casting strategy again. 79d6599 will break but d5ea425 will pass.

I'm also updating all code snippets to use the new Elemental 2 API in our docs.

**Comment by mbarkley** [comment]
@tiagobento Okay. Let me know when you think it's ready.

**Comment by tiagobento** [comment]
@mbarkley It's ready to merge.

**Pull request merged**